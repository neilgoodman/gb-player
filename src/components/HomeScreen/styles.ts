import { StyleSheet } from 'react-native';
import styleGuide from '../../styleGuide';

export default StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: styleGuide.colors.secondaryBackground,
    alignItems: 'center',
    justifyContent: 'center',
  },
  text: {
    color: styleGuide.colors.white,
  },
});
