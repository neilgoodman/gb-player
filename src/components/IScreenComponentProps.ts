import { NavigationRoute, NavigationScreenProp } from 'react-navigation';

export default interface IScreenComponentProps {
  navigation: NavigationScreenProp<NavigationRoute>;
};
