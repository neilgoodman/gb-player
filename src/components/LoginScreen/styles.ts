import { StyleSheet } from 'react-native';
import styleGuide from '../../styleGuide';

export default StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: styleGuide.colors.secondaryBackground,
    paddingTop: 40,
    paddingRight: 20,
    paddingLeft: 20,
  },
  text: {
    fontSize: 16,
    color: styleGuide.colors.white,
    marginBottom: 20,
  },
});
