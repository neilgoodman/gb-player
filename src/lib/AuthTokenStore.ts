import { SecureStore } from 'expo';

const AUTH_TOKEN_STORE_KEY = 'gb-player-auth-token';

/**
 * Utility for storing the Giant Bomb API key locally on the device.
 */
export default class AUthTokenStore {
  /**
   * Get the Giant Bomb API key.
   *
   * Returns empty string if there is no API key available.
   */
  public static async getAsync(): Promise<string> {
    const authToken = await SecureStore.getItemAsync(AUTH_TOKEN_STORE_KEY);
    return authToken == null ? '' : authToken;
  }

  /**
   * Stores the given API key to the device.
   * @param token The Giant Bomb API key to store.
   */
  public static async setAsync(token: string): Promise<void> {
    await SecureStore.setItemAsync(AUTH_TOKEN_STORE_KEY, token);
  }

  /**
   * Clears the API key from the device.
   */
  public static async clearAsync(): Promise<void> {
    await SecureStore.deleteItemAsync(AUTH_TOKEN_STORE_KEY);
  }
}
