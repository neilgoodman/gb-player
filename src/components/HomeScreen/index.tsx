import React from 'react';
import { Text, View } from 'react-native';
import AuthTokenStore from '../../lib/AuthTokenStore';
import Button from '../Button';
import IScreenComponentProps from '../IScreenComponentProps';
import LoginScreen from '../LoginScreen';
import styles from './styles';

export interface IState {
  authToken?: string;
}

export default class HomeScreen extends React.Component<
  IScreenComponentProps,
  IState
> {
  public static ROUTE = 'HOME_SCREEN_ROUTE';

  constructor(props) {
    super(props);
    this.state = {
      authToken: '',
    };
  }

  public async componentDidMount() {
    const authToken = await AuthTokenStore.getAsync();

    this.setState({
      authToken,
    });

    if (authToken === '') {
      this.props.navigation.navigate(LoginScreen.ROUTE);
    }
  }

  public onPress = (): void => {
    AuthTokenStore.clearAsync();
  };

  public render() {
    return (
      <View style={styles.container}>
        <Text style={styles.text}>Home Screen {this.state.authToken}</Text>
        <Button onPress={this.onPress}>Clear Auth Token</Button>
      </View>
    );
  }
}
