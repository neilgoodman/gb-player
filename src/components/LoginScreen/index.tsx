import React from 'react';
import { Linking, Text, View } from 'react-native';
import AuthTokenStore from '../../lib/AuthTokenStore';
import { Client, GetResultPoller } from '../../lib/giant-bomb-api-client/auth';
import styleGuide from '../../styleGuide';
import Button from '../Button';
import HomeScreen from '../HomeScreen';
import IScreenComponentProps from '../IScreenComponentProps';
import styles from './styles';

export interface IState {
  loginEndpoint: string;
  hasLoaded: boolean;
  getResultPoller: GetResultPoller | null;
  regCode: string;
}

export default class LoginScreen extends React.Component<
  IScreenComponentProps,
  IState
> {
  public static ROUTE = 'LOGIN_SCREEN_ROUTE';
  private _client: Client;

  constructor(props: IScreenComponentProps) {
    super(props);
    this.state = {
      loginEndpoint: Client.LOGIN_ENDPOINT,
      hasLoaded: false,
      getResultPoller: null,
      regCode: '',
    };
    this._client = new Client();
  }

  public async componentWillMount() {
    const getCodeResult = await this._client.getCodeAsync();
    this.setState({
      regCode: getCodeResult.regCode,
      getResultPoller: new GetResultPoller(getCodeResult),
      hasLoaded: true,
    });
  }

  public componentWillUnmount() {
    const { getResultPoller } = this.state;

    if (getResultPoller) {
      getResultPoller.stop();
    }
  }

  public async componentDidUpdate(
    prevProps: IScreenComponentProps,
    prevState: IState
  ) {
    const { getResultPoller } = this.state;
    if (getResultPoller && !getResultPoller.hasStarted()) {
      const getResultResult = await getResultPoller.startAsync();

      if (getResultResult !== null) {
        await AuthTokenStore.setAsync(getResultResult.regToken);
        this.props.navigation.navigate(HomeScreen.ROUTE);
      }
    }
  }

  public openGiantBombOnPress = () => {
    Linking.openURL(this.state.loginEndpoint);
  };

  public render() {
    return <View style={styles.container}>{this._renderLoginDetails()}</View>;
  }

  private _renderLoginDetails(): JSX.Element | null {
    const { loginEndpoint, hasLoaded, regCode } = this.state;

    if (!hasLoaded) {
      return null;
    }

    return (
      <View>
        <Text style={styles.text}>
          Visit {loginEndpoint} and enter the code below to login.
        </Text>
        <Text style={styles.text}>{regCode}</Text>
        <Button onPress={this.openGiantBombOnPress}>Go to Giant Bomb</Button>
      </View>
    );
  }
}
