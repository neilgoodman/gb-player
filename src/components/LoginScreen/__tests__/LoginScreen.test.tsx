jest.mock('../../../lib/giant-bomb-api-client/auth');
jest.mock('../../../lib/AuthTokenStore');
jest.mock('Linking');

import React from 'react';
import { Linking } from 'react-native';
import renderer from 'react-test-renderer';
import LoginScreen from '../';
import AuthTokenStore from '../../../lib/AuthTokenStore';
import {
  Client,
  GetCode,
  GetResultPoller,
} from '../../../lib/giant-bomb-api-client/auth';
import HomeScreen from '../../HomeScreen';
import IScreenComponentProps from '../../IScreenComponentProps';

describe('LoginScreen', () => {
  let props: IScreenComponentProps;
  const TEST_REG_CODE = 'C12345';
  const getCode = {
    status: 'success',
    regCode: TEST_REG_CODE,
    retryInterval: 1,
    retryDuration: 1,
  };
  const TEST_REG_TOKEN = 'D12345';

  beforeEach(() => {
    // @ts-ignore: jest mock
    props = {
      navigation: {
        navigate: jest.fn(),
      },
    };
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  describe('render', () => {
    it('should render correctly when given all props', () => {
      const rendered = renderer
        .create(<LoginScreen navigation={props.navigation} />)
        .toJSON();

      expect(rendered).toMatchSnapshot();
    });

    it('should render correctly when state.hasLoaded and state.regCode are set', () => {
      const testRenderer = renderer.create(
        <LoginScreen navigation={props.navigation} />
      );

      const instance = testRenderer.getInstance();
      // @ts-ignore: test renderer
      instance.setState({
        hasLoaded: true,
        regCode: TEST_REG_CODE,
      });

      const rendered = testRenderer.toJSON();

      expect(rendered).toMatchSnapshot();
    });
  });

  describe('componentWillMount', () => {
    it('should update state with results from auth/Client.getCodeAsync', async () => {
      expect.assertions(3);

      const loginScreen = new LoginScreen(props);

      // @ts-ignore: jest mock
      loginScreen._client.getCodeAsync.mockResolvedValueOnce(getCode);
      loginScreen.setState = jest.fn();

      await loginScreen.componentWillMount();

      // @ts-ignore: jest mock
      const setStateCall = loginScreen.setState.mock.calls[0][0];

      expect(setStateCall.regCode).toBe(TEST_REG_CODE);
      expect(setStateCall.getResultPoller).toBeInstanceOf(GetResultPoller);
      expect(setStateCall.hasLoaded).toBe(true);
    });
  });

  describe('componentWillUnmount', () => {
    it('should call state.getResultPoller.stop if set', () => {
      const loginScreen = new LoginScreen(props);

      // @ts-ignore: jest mock
      loginScreen.state = {
        getResultPoller: {
          stop: jest.fn(),
        },
      };
      loginScreen.componentWillUnmount();

      // @ts-ignore: jest
      expect(loginScreen.state.getResultPoller.stop).toHaveBeenCalled();
    });

    it("should't throw an error if state.getResultPoller is not set", () => {
      const loginScreen = new LoginScreen(props);

      expect(() => loginScreen.componentWillUnmount()).not.toThrow();
    });
  });

  describe('componentDidUpdate', () => {
    it('should fetch the API key, store it, and navigate to HomeScreen if state.getResultPoller is set', async () => {
      expect.assertions(2);

      const loginScreen = new LoginScreen(props);

      // @ts-ignore: jest mock
      loginScreen.state = {
        getResultPoller: {
          hasStarted: jest.fn().mockReturnValueOnce(false),
          startAsync: jest
            .fn()
            .mockResolvedValueOnce({ regToken: TEST_REG_TOKEN }),
        },
      };

      // @ts-ignore: jest
      await loginScreen.componentDidUpdate(props, {});

      expect(AuthTokenStore.setAsync).toHaveBeenCalledWith(TEST_REG_TOKEN);
      expect(props.navigation.navigate).toHaveBeenCalledWith(HomeScreen.ROUTE);
    });

    it("shouldn't store API key and navigate to HomeScreen if regToken was null", async () => {
      expect.assertions(2);

      const loginScreen = new LoginScreen(props);

      // @ts-ignore: jest mock
      loginScreen.state = {
        getResultPoller: {
          hasStarted: jest.fn().mockReturnValueOnce(false),
          startAsync: jest.fn().mockResolvedValueOnce(null),
        },
      };

      // @ts-ignore: jest
      await loginScreen.componentDidUpdate(props, {});

      expect(AuthTokenStore.setAsync).not.toHaveBeenCalledWith(TEST_REG_TOKEN);
      expect(props.navigation.navigate).not.toHaveBeenCalledWith(
        HomeScreen.ROUTE
      );
    });

    it("shouldn't store API key and navigate to HomeScreen if getResultPoller hasStarted returns true", async () => {
      expect.assertions(2);

      const loginScreen = new LoginScreen(props);

      // @ts-ignore: jest mock
      loginScreen.state = {
        getResultPoller: {
          hasStarted: jest.fn().mockReturnValueOnce(true),
          startAsync: jest.fn().mockResolvedValueOnce(null),
        },
      };

      // @ts-ignore: jest
      await loginScreen.componentDidUpdate(props, {});

      expect(AuthTokenStore.setAsync).not.toHaveBeenCalledWith(TEST_REG_TOKEN);
      expect(props.navigation.navigate).not.toHaveBeenCalledWith(
        HomeScreen.ROUTE
      );
    });
  });

  describe('openGiantBombOnPress', () => {
    it('should deep link to the Giant Bomb login endpoint', () => {
      const loginScreen = new LoginScreen(props);

      loginScreen.openGiantBombOnPress();

      // @ts-ignore: jest mock
      expect(Linking.openURL.mock.calls[0][0]).toBe(
        'https://www.giantbomb.com/app/gb-player/activate'
      );
    });
  });
});
