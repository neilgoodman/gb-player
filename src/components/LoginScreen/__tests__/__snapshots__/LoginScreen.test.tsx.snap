// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen render should render correctly when given all props 1`] = `
<View
  style={
    Object {
      "backgroundColor": "rgb(44,46,48)",
      "flex": 1,
      "paddingLeft": 20,
      "paddingRight": 20,
      "paddingTop": 40,
    }
  }
/>
`;

exports[`LoginScreen render should render correctly when state.hasLoaded and state.regCode are set 1`] = `
<View
  style={
    Object {
      "backgroundColor": "rgb(44,46,48)",
      "flex": 1,
      "paddingLeft": 20,
      "paddingRight": 20,
      "paddingTop": 40,
    }
  }
>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "rgb(255,255,255)",
          "fontSize": 16,
          "marginBottom": 20,
        }
      }
    >
      Visit 
      https://www.giantbomb.com/app/gb-player/activate
       and enter the code below to login.
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "rgb(255,255,255)",
          "fontSize": 16,
          "marginBottom": 20,
        }
      }
    >
      C12345
    </Text>
    <View
      accessibilityComponentType="button"
      accessibilityLabel={undefined}
      accessibilityTraits="button"
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#007aff",
              "fontSize": 17,
              "fontWeight": "500",
              "textAlign": "center",
            },
            null,
            Array [
              Object {
                "backgroundColor": "rgb(81,84,88)",
                "color": "rgb(255,255,255)",
                "marginBottom": 40,
                "paddingBottom": 10,
                "paddingTop": 10,
                "width": 710,
              },
            ],
            null,
          ]
        }
      >
        Go to Giant Bomb
      </Text>
    </View>
  </View>
</View>
`;
